tests/unit/python/foglamp/plugins/south/coap_listen/test_coap_listen.py

```diff
@@ -0,0 +1,224 @@
+# -*- coding: utf-8 -*-
+
+# FOGLAMP_BEGIN
+# See: http://foglamp.readthedocs.io/
+# FOGLAMP_END
+
+"""Unit test for foglamp.plugins.south.coap_listen.coap_listen.py"""
+import copy
+import json
+import pytest
+import asyncio
+from unittest import mock
+from unittest.mock import call, patch
+from aiohttp.test_utils import make_mocked_request
+from aiohttp.streams import StreamReader
+from multidict import CIMultiDict
+import cbor2
+from aiocoap import message, numbers
+
+from foglamp.plugins.south.coap_listen import coap_listen
+from foglamp.plugins.south.coap_listen.coap_listen import CoAPIngest
+from foglamp.plugins.south.coap_listen.coap_listen import Ingest
+
+
+__author__ = "Amarendra K Sinha"
+__copyright__ = "Copyright (c) 2017 OSIsoft, LLC"
+__license__ = "Apache 2.0"
+__version__ = "${VERSION}"
+
+
+def _DEFAULT_CONFIG(unused_port):
+    return {
+            'plugin': {
+                 'description': 'Python module name of the plugin to load',
+                 'type': 'string',
+                 'default': 'coap_listen'
+            },
+            'port': {
+                'description': 'Port to listen on',
+                'type': 'integer',
+                'default': str(unused_port),
+            },
+            'uri': {
+                'description': 'URI to accept data on',
+                'type': 'string',
+                'default': 'sensor-values',
+            }
+    }
+
+_NEW_CONFIG = {
+    'plugin': {
+         'description': 'Python module name of the plugin to load',
+         'type': 'string',
+         'default': 'coap_listen'
+    },
+    'port': {
+        'description': 'Port to listen on',
+        'type': 'integer',
+        'default': '1234',
+    },
+    'uri': {
+        'description': 'URI to accept data on',
+        'type': 'string',
+        'default': 'sensor-values',
+    }
+}
+
+
+def test_plugin_info():
+    assert coap_listen.plugin_info() == {
+            'name': 'CoAP Server',
+            'version': '1.0',
+            'mode': 'async',
+            'type': 'south',
+            'interface': '1.0',
+            'config': _DEFAULT_CONFIG(5683)
+    }
+
+def test_plugin_init(unused_port):
+    config = _DEFAULT_CONFIG(unused_port())
+    assert coap_listen.plugin_init(config) == config
+
+def test_plugin_start(mocker, unused_port):
+    # GIVEN
+    config = copy.deepcopy(_DEFAULT_CONFIG(unused_port()))
+    config['port']['value'] = config['port']['default']
+    config['uri']['value'] = config['uri']['default']
+
+    log_exception = mocker.patch.object(coap_listen._LOGGER, "exception")
+    log_info = mocker.patch.object(coap_listen._LOGGER, "info")
+    assert coap_listen.aiocoap_ctx is None
+
+    # WHEN
+    loop = asyncio.get_event_loop()
+    loop.call_soon(coap_listen.plugin_start(config))
+
+    # THEN
+    assert coap_listen.aiocoap_ctx is not None
```

Comment:
This test logic shouldn't be in test_plugin_start but rather in a test for _start_aiocoap.

Reply:
The point is - we should keep mocking at bare minimum. This will help us in finding fault in code at the earliest and conveniently.